File: src/test/java/graphql/kickstart/tools/util/BiMapTest.java
package graphql.kickstart.tools.util;

import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class BiMapTest {
    private BiMap<String, Integer> bimap;

    @Before
    public void setUp() {
        bimap = BiMap.create();
    }

    @Test(expected = IllegalArgumentException.class)
    public void testDoesNotPermitDuplicateValues() {
        bimap.put("foo", 1);
        bimap.put("bar", 1);
    }

    @Test
    @SuppressWarnings("OverwrittenKey")
    public void testPutSameKeySameValue() {
        bimap.put("foo", 1);

        // should not fail
        bimap.put("foo", 1);
    }

    @Test
    @SuppressWarnings({"ConstantConditions"})
    public void testPutSameKeyNewValue() {
        bimap.put("foo", 1);
        bimap.put("bar", 2);

        int old = bimap.put("bar", 3);

        // old value should have been returned
        assertEquals(2, old);

        // inverse should have correct keys
        Set<Integer> expected = new HashSet<>(Arrays.asList(1, 3));
        assertEquals(expected, bimap.values());
        assertEquals(expected, bimap.inverse().keySet());
    }

    @Test
    public void testInverse() {
        bimap.put("foo", 1);
        bimap.put("bar", 2);
        bimap.put("baz", 3);

        BiMap<Integer, String> inverse = bimap.inverse();
        assertTrue(inverse.containsKey(1));
        assertTrue(inverse.containsKey(2));
        assertTrue(inverse.containsKey(3));

        assertThat(inverse.get(1), is("foo"));
        assertThat(inverse.get(2), is("bar"));
        assertThat(inverse.get(3), is("baz"));
    }

    @Test
    public void testValues() {
        bimap.put("foo", 1);
        bimap.put("bar", 2);
        bimap.put("baz", 3);

        Set<Integer> values = bimap.values();
        Set<Integer> expected = new HashSet<>(Arrays.asList(1, 2, 3));
        assertThat(values, is(expected));
    }
}

File: src/test/kotlin/graphql/kickstart/tools/util/BiMapTest.kt
package graphql.kickstart.tools.util

import graphql.kickstart.tools.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import java.util.*

class BiMapTest {

    lateinit var bimap: BiMap<String, Int>

    @Before
    fun setup() {
        bimap = BiMap.create()
    }

    @Test(expected = IllegalArgumentException::class)
    fun `should not permit duplicate values`() {
        bimap["foo"] = 1
        bimap["bar"] = 1
    }

    @Test
    fun `should allow same key same value`() {
        bimap["foo"] = 1

        // should not fail
        bimap["foo"] = 1
    }

    @Test
    fun `should allow same key new value`() {
        bimap["foo"] = 1
        bimap["bar"] = 2
        val old = bimap.put("bar", 3)!!

        // old value should have been returned
        assertEquals(2, old.toLong())

        // inverse should have correct keys
        val expected: Set<Int> = HashSet(Arrays.asList(1, 3))
        assertEquals(expected, bimap.values)
        assertEquals(expected, bimap.inverse().keys)
    }

    @Test
    fun `should provide inverse view`() {
        bimap["foo"] = 1
        bimap["bar"] = 2
        bimap["baz"] = 3
        val inverse = bimap.inverse()
        assertTrue(inverse.containsKey(1))
        assertTrue(inverse.containsKey(2))
        assertTrue(inverse.containsKey(3))
        assertEquals(inverse[1], "foo")
        assertEquals(inverse[2], "bar")
        assertEquals(inverse[3], "baz")
    }

    @Test
    fun `should return correct values`() {
        bimap["foo"] = 1
        bimap["bar"] = 2
        bimap["baz"] = 3
        val values = bimap.values
        val expected = setOf(1, 2, 3)
        assertEquals(expected, values)
    }
}
